### ISSUE TITLE
including nsfw small from oisd failed

### ISSUE THREAD

**Issue by embolon** [lynxthecat/adblock-lean#87]
I try to add the nsfw list from oisd but it failed to start. oisd version is "# Version: 202412270213"

I've replaced the blocklist_urls with "https://nsfw-small.oisd.nl/domainswild2"
and get this warning and error:

Downloading, checking and sanitizing raw blocklist part from: https://nsfw-small.oisd.nl/domainswild2.
Warning: Rogue element: 'adult' identified originating in blocklist file from: https://nsfw-small.oisd.nl/domainswild2.
Skipping file and continuing.

Error: Failed to generate preprocessed blocklist file with at least one entry.

Restoring saved blocklist file.

Error: No previous blocklist file found.

Error: Failed to restore saved blocklist.

Stopping adblock-lean.
Removing any adblock-lean blocklist files in /tmp/dnsmasq.d.

Restarting dnsmasq.

Waiting for dnsmasq initialization.
Restart of dnsmasq completed.

Stopped adblock-lean.

**Comment by friendly-bits** [comment]
This list should work with advlock-lean, except it has an entry `adult` which is not a valid domain. adblock-lean correctly identifies it as a rogue element. When encountering a rogue element, adblock-lean disqualifies the list because besides mistakes on the part of list maintainers, a rogue element may show up in lists which try to hijack or misconfigure dns resolution, or when a download error occurs and you receive a corrupted list. adblock-lean has no way to tell these situations apart, so it simply disqualifies the list.

The solution here is to contact the list maintainer and ask the to remove the invalid entry.

**Comment by friendly-bits** [comment]
Actually, looking at the non-wild NSFW list from oisd, I realized that 'adult's is meant as a TLD. So the list essentially wants to block the whole TLD 'adult'. Now I'm not sure how this should be handled by adblock-lean. Perhaps we should add an option to allow TLD blocking and when set, use a regex which allows that? @lynxthecat @Wizballs @dave14305

**Comment by lynxthecat** [comment]
Is there a dnsmasq compatible syntax for that kind of block?

**Comment by friendly-bits** [comment]
> Is there a dnsmasq compatible syntax for that kind of block?

The supposedly equivalent oisd [dnsmasq-formatted list](https://nsfw.oisd.nl/dnsmasq2) has this entry:
`local=/adult/`

So I suppose this is the syntax.

**Comment by friendly-bits** [comment]
Giving this some more thought, I don't see many downsides in simply allowing TLD blocking without any extra options. The only actual downside is that this could potentially allow a partially-downloaded list in, but I think chances for that are fairly slim and perhaps the inconvenience for the user associated with figuring out that they need to change an option (and the support effort on our side) outweigh this potential disadvantage. What do you think?

**Comment by lynxthecat** [comment]
Makes sense to me. And don't we still check that the download completed anyway?

**Comment by friendly-bits** [comment]
> don't we still check that the download completed anyway?

What we do is check the STDERR output of uclient-fetch and look for the "Download completed" string in it:
https://github.com/lynxthecat/adblock-lean/blob/c1656d1b2335c8a38a4b497da2f2532968ee59a1/adblock-lean#L1171

This does not guarantee that in all cases, the download completed successfully because uclient-fetch may not necessarily know whether it received all the data and whether the data is intact or not. I don't know enough about TCP to elaborate more on possible failure modes, just saying that "Download completed" does not guarantee a complete and intact transfer, to my understanding.

We do check the resulting list with dnsmasq --test, which at least makes it unlikely to that upon actually loading the blocklist we will break dnsmasq potentially causing it to go into restart loop. This still does not guarantee the the all blocklist parts are intact.

However, I think that the unaddressed download failure modes should be rare enough that we can probably ignore them.

**Comment by friendly-bits** [comment]
This should be solved in current 'master' branch which adds support for blocking TLD's.
